File: day11.py
#!/usr/bin/env python3


def occupied(seats, r, c, search):
    count = 0
    for i in range(-1,2):
        for j in range(-1,2):
            if i == 0 and j == 0:
                continue
            rr = r + i
            cc = c + j
            if search:
                while 0 <= rr < len(seats) and 0 <= cc < len(seats[0]) and seats[rr][cc] == '.':
                    rr += i
                    cc += j
            if 0 <= rr < len(seats) and 0 <= cc < len(seats[0]) and seats[rr][cc] == '#':
                count += 1
    return count


def part1(seats, search, num):
    old = None
    new = [[x for x in seat] for seat in seats]
    while new != old:
        old = [[x for x in seat] for seat in new]
        for r, row in enumerate(old):
            for c, seat in enumerate(row):
                if seat == 'L' and occupied(old, r, c, search) == 0:
                    seat = '#'
                elif seat == '#' and occupied(old, r, c, search) >= num:
                    seat = 'L'
                new[r][c] = seat
    return sum([row.count('#') for row in new])


if __name__ == "__main__":
    with open("day11.txt") as f:
        problem = [[char for char in line] for line in f.readlines()]

    sample = """L.LL.LL.LL
LLLLLLL.LL
L.L.L..L..
LLLL.LL.LL
L.LL.LL.LL
L.LLLLL.LL
..L.L.....
LLLLLLLLLL
L.LLLLLL.L
L.LLLLL.LL""".split('\n')

    sample = [[char for char in word] for word in sample]
    assert part1(sample, False, 4) == 37
    ans1 = part1(problem, False, 4)
    print(ans1)
    assert ans1 == 2406

    assert part1(sample, True, 5) == 26
    ans2 = part1(problem, True, 5)
    print(ans2)
    assert ans2 == 2149

File: src/day11.rs
use std::fs;

fn occupied(seats: &[Vec<char>], r: usize, c:usize, search: bool) -> usize {
    let mut count = 0;
    for ii in 0..3 {
        for jj in 0..3 {
            let i:isize = ii - 1;
            let j:isize = jj - 1;
            if i == j && j == 0 {
                continue;
            }
            let mut rr: isize = r as isize + i;
            let mut cc: isize = c as isize + j;
            if search {
                while 0 <= rr && rr < seats.len() as isize && cc < seats[0].len() as isize && cc >= 0 && seats[rr as usize][cc as usize] == '.' {
                    rr += i;
                    cc += j;
                }
            }
            if rr >= 0 && cc >= 0 && rr < (seats.len() as isize) && cc < (seats[0].len() as isize) && seats[rr as usize][cc as usize] == '#' {
                count += 1;
            }
        }
    }
    count
}

fn part1(seats: &[Vec<char>], search: bool, num: usize) -> usize {
    let mut old: Vec<Vec<char>> = vec![vec!['_'; seats[0].len()]; seats.len()];
    let mut new: Vec<Vec<char>> = seats.iter().map(|x| x.to_vec()).collect();
    while new != old {
        old = new.iter().map(|x| x.to_vec()).collect();
        for (r, row) in old.iter().enumerate() {
            for (c, seat) in row.iter().enumerate() {
                let mut new_seat = *seat;
                if *seat == 'L' && occupied(&old, r, c, search) == 0 {
                    new_seat = '#';
                }
                if *seat == '#' && occupied(&old, r, c, search) >= num {
                    new_seat = 'L';
                }
                new[r][c] = new_seat;
            }
        }
    }
    return new.iter().map(|r| r.iter().filter(|s| **s == '#').count()).sum();
}

pub fn day11() {
    let sample: Vec<&str> = "L.LL.LL.LL
LLLLLLL.LL
L.L.L..L..
LLLL.LL.LL
L.LL.LL.LL
L.LLLLL.LL
..L.L.....
LLLLLLLLLL
L.LLLLLL.L
L.LLLLL.LL".lines().collect();
    let sample_vec: Vec<Vec<char>> = sample.iter().map(|x| (*x).chars().collect::<Vec<char>>()).collect();
    let contents = fs::read_to_string("day11.txt").unwrap();
    let input_vec: Vec<Vec<char>> = contents.lines().collect::<Vec<&str>>().iter().map(|x| x.chars().collect::<Vec<char>>()).collect();

    assert_eq!(part1(&sample_vec, false, 4), 37);
    let ans1 = part1(&input_vec, false, 4);
    println!("{}", ans1);
    assert_eq!(ans1, 2406);

    assert_eq!(part1(&sample_vec, true, 5), 26);
    let ans1 = part1(&input_vec, true, 5);
    println!("{}", ans1);
    assert_eq!(ans1, 2149);
}